mayuctl/list.go

```diff
@@ -2,44 +2,144 @@ package main
 	"github.com/spf13/cobra"
 )
 
+const (
+	defaultListFields = "ip,serial,profile,ipmiaddr,providerid,metadata,coreos,state,lastboot"
+	timestampFormat   = "2006-01-02 15:04:05"
+)
+
+type ListFlags struct {
+	Fields   string
+	NoLegend bool
+}
+
+type hostToField func(host *hostmgr.Host) string
+
 var (
 	listCmd = &cobra.Command{
-		Use:   "list",
+		Use:   "list [--no-legend] [--fields]",
 		Short: "List machines.",
-		Long:  "List machines.",
-		Run:   listRun,
+		Long: `List the state of all machines in a Mayu datacenter.
+
+For easily parsable output, you can remove the column headers:
+	mayuctl list --no-legend
+
+Or, choose the columns to display:
+	mayuctl list --fields=ip,machineid,ipmiaddr`,
+		Run: listRun,
 	}
-)
 
-const (
-	listHeader      = "IP | Serial | Profile | IPMI Address | ProviderId | Fleet | CoreOS | State | Last Boot"
-	listScheme      = "%s | %s | %s | %s | %s | %s | %s | %s | %s"
-	timestampFormat = "2006-01-02 15:04:05"
+	listFields = map[string]hostToField{
+		"ip": func(host *hostmgr.Host) string {
+			if host.InternalAddr == nil {
+				return "-"
+			}
+			return host.InternalAddr.String()
+		},
+		"serial": func(host *hostmgr.Host) string {
+			return host.Serial
+		},
+		"profile": func(host *hostmgr.Host) string {
+			if host.Profile == "" {
+				return "-"
+			}
+			return host.Profile
+		},
+		"ipmiaddr": func(host *hostmgr.Host) string {
+			if host.IPMIAddr == nil {
+				return "-"
+			}
+			return host.IPMIAddr.String()
+		},
+		"providerid": func(host *hostmgr.Host) string {
+			if host.ProviderId == "" {
+				return "-"
+			}
+			return host.ProviderId
+		},
+		"metadata": func(host *hostmgr.Host) string {
+			return host.FleetMetadata.String()
+		},
+		"coreos": func(host *hostmgr.Host) string {
+			return host.CoreOSVersion
+		},
+		"yochu": func(host *hostmgr.Host) string {
+			return host.YochuVersion
+		},
+		"fleet": func(host *hostmgr.Host) string {
+			return host.FleetVersion
+		},
+		"etcd": func(host *hostmgr.Host) string {
+			return host.EtcdVersion
+		},
+		"docker": func(host *hostmgr.Host) string {
+			return host.DockerVersion
+		},
+		"machineid": func(host *hostmgr.Host) string {
+			return host.MachineID
+		},
+		"state": func(host *hostmgr.Host) string {
+			return hostmgr.HostStateMap()[host.State]
+		},
+		"lastboot": func(host *hostmgr.Host) string {
+			return host.LastBoot.Format(timestampFormat)
+		},
+	}
+
+	listFlags = &ListFlags{}
 )
 
+func init() {
+	listCmd.PersistentFlags().BoolVar(&listFlags.NoLegend, "no-legend", false, "Do not print a legend (column headers)")
+	listCmd.PersistentFlags().StringVar(&listFlags.Fields, "fields", defaultListFields, fmt.Sprintf("Columns to print for each Machine. Valid fields are %q", strings.Join(hostToFieldKeys(listFields), ",")))
+}
+
 func listRun(cmd *cobra.Command, args []string) {
+	if listFlags.Fields == "" {
+		fmt.Printf("You have to define fields for the output. This can't be empty.\n\nUsage: %s\n", cmd.Usage())
```

Comment:
Better use this message `Invalid fields parameter. Please, choose valid fields `ip,serial,...`